File: src/utils/fixtureGenerator.ts
import { type League, type Match } from '../types/game';

export function generateFixtures(leagues: League[], clubIdsByLeague: Record<string, string[]>, season: number): Match[] {
  const allMatches: Match[] = [];

  leagues.forEach(league => {
    const clubIds = [...(clubIdsByLeague[league.id] || [])];
    if (clubIds.length < 2) return;

    // Ensure even number of clubs
    if (clubIds.length % 2 !== 0) clubIds.push('BYE');

    const numClubs = clubIds.length;
    const rounds = (numClubs - 1) * 2; // Home and Away
    const matchesPerRound = numClubs / 2;

    for (let round = 0; round < rounds; round++) {
      const weekNum = round + 1;
      if (weekNum > 38) break; // Hard limit to 38 weeks for now

      for (let i = 0; i < matchesPerRound; i++) {
        const homeIdx = i;
        const awayIdx = numClubs - 1 - i;

        const homeId = clubIds[homeIdx];
        const awayId = clubIds[awayIdx];

        if (homeId === 'BYE' || awayId === 'BYE') continue;

        allMatches.push({
          id: `m-${league.id}-${round}-${homeId}-${awayId}-${season}`,
          homeClubId: round % 2 === 0 ? homeId : awayId,
          awayClubId: round % 2 === 0 ? awayId : homeId,
          homeScore: 0,
          awayScore: 0,
          played: false,
          leagueId: league.id,
          season,
          week: weekNum,
          events: []
        });
      }

      // Rotate clubs for Round Robin
      const last = clubIds.pop()!;
      clubIds.splice(1, 0, last);
    }
  });

  return allMatches;
}

File: chairman/backend/simulation/fixture_generator.py
from game.models import League, Club, Match

def generate_fixtures(season: int) -> int:
    leagues = League.objects.all()
    fixtures_created = 0

    for league in leagues:
        club_ids = list(Club.objects.filter(league=league).values_list('id', flat=True))
        if len(club_ids) < 2:
            continue

        # Ensure even number of clubs
        if len(club_ids) % 2 != 0:
            club_ids.append('BYE')

        num_clubs = len(club_ids)
        rounds = (num_clubs - 1) * 2 # Home and Away
        matches_per_round = num_clubs // 2

        for round_idx in range(rounds):
            week_num = round_idx + 1
            if week_num > 38:
                break

            for i in range(matches_per_round):
                home_idx = i
                away_idx = num_clubs - 1 - i

                home_id = club_ids[home_idx]
                away_id = club_ids[away_idx]

                if home_id == 'BYE' or away_id == 'BYE':
                    continue

                # Alternate home/away based on round % 2
                final_home_id = home_id if round_idx % 2 == 0 else away_id
                final_away_id = away_id if round_idx % 2 == 0 else home_id

                Match.objects.create(
                    league=league,
                    home_club_id=final_home_id,
                    away_club_id=final_away_id,
                    season=season,
                    week=week_num,
                    played=False,
                    home_score=0,
                    away_score=0,
                    events=[],
                    player_ratings={}
                )
                fixtures_created += 1

            # Rotate clubs for Round Robin (fixed the first element, rotate others)
            last = club_ids.pop()
            club_ids.insert(1, last)

    return fixtures_created
